File: software/examples/c/example_callback.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_dual_button_v2.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your Dual Button Bricklet 2.0

// Callback function for state changed callback
void cb_state_changed(uint8_t button_l, uint8_t button_r, uint8_t led_l, uint8_t led_r,
                      void *user_data) {
	(void)led_l; (void)led_r; (void)user_data; // avoid unused parameter warning

	if(button_l == DUAL_BUTTON_V2_BUTTON_STATE_PRESSED) {
		printf("Left button pressed\n");
	} else {
		printf("Left button released\n");
	}

	if(button_r == DUAL_BUTTON_V2_BUTTON_STATE_PRESSED) {
		printf("Right button pressed\n");
	} else {
		printf("Right button released\n");
	}

	printf("\n");
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	DualButtonV2 db;
	dual_button_v2_create(&db, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register state changed callback to function cb_state_changed
	dual_button_v2_register_callback(&db,
	                                 DUAL_BUTTON_V2_CALLBACK_STATE_CHANGED,
	                                 (void *)cb_state_changed,
	                                 NULL);

	printf("Press key to exit\n");
	getchar();
	dual_button_v2_destroy(&db);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{dual_button_v2_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Dual Button Bricklet 2.0.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let db = DualButtonV2Bricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let state_changed_receiver = db.get_state_changed_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `db` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for state_changed in state_changed_receiver {
            if state_changed.button_l == DUAL_BUTTON_V2_BRICKLET_BUTTON_STATE_PRESSED {
                println!("Left Button: Pressed");
            } else if state_changed.button_l == DUAL_BUTTON_V2_BRICKLET_BUTTON_STATE_RELEASED {
                println!("Left Button: Released");
            }

            if state_changed.button_r == DUAL_BUTTON_V2_BRICKLET_BUTTON_STATE_PRESSED {
                println!("Right Button: Pressed");
            } else if state_changed.button_r == DUAL_BUTTON_V2_BRICKLET_BUTTON_STATE_RELEASED {
                println!("Right Button: Released");
            }

            println!();
        }
    });

    // Enable state changed callback
    db.set_state_changed_callback_configuration(true).recv()?;

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
